core/kotlinx-coroutines-core/src/internal/Exceptions.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.experimental.internal
+
+import kotlinx.coroutines.experimental.*
+import kotlin.coroutines.experimental.*
+import kotlin.coroutines.experimental.jvm.internal.*
+
+internal actual fun <E : Throwable> recoverStackTrace(exception: E): E {
+    if (!DEBUG || exception is CancellationException) {
+        return exception
+    }
+
+    return tryWrapException(exception) ?: exception
+}
+
+internal actual fun <E : Throwable> recoverStackTrace(exception: E, continuation: Continuation<*>): E {
+    if (!DEBUG || exception is CancellationException || continuation !is CoroutineImpl) {
```

Comment:
`continuation !is CoroutineImpl` is too restrictive for experimental coroutines. `CancellableContinuationImpl` shall be supported, too. It is needed for basic example that use `Deferred.await()`, for example.